crates/ruff_db/src/system/os.rs

```diff
@@ -75,11 +89,154 @@ impl System for OsSystem {
         &self,
         path: &SystemPath,
     ) -> Result<Box<dyn Iterator<Item = Result<DirectoryEntry>>>> {
-        Ok(Box::new(
-            path.as_utf8_path()
-                .read_dir_utf8()?
-                .map(|res| res.map(DirectoryEntry::from)),
-        ))
+        Ok(Box::new(path.as_utf8_path().read_dir_utf8()?.map(|res| {
+            let res = res?;
+
+            let file_type = res.file_type()?;
+            Ok(DirectoryEntry {
+                path: SystemPathBuf::from_utf8_path_buf(res.into_path()),
+                file_type: file_type.into(),
+            })
+        })))
+    }
+}
+
+#[derive(Debug)]
+struct OsDirectoryWalker;
+
+impl DirectoryWalker for OsDirectoryWalker {
+    fn walk(
+        &self,
+        visitor_builder: &mut dyn WalkDirectoryVisitorBuilder,
+        configuration: WalkDirectoryConfiguration,
+    ) {
+        let WalkDirectoryConfiguration {
+            paths,
+            ignore_hidden: hidden,
+            standard_filters,
+        } = configuration;
+
+        let Some((first, additional)) = paths.split_first() else {
+            return;
+        };
+
+        let mut builder = ignore::WalkBuilder::new(first.as_std_path());
+
+        builder.standard_filters(standard_filters);
+        builder.hidden(hidden);
+
+        for additional_path in additional {
+            builder.add(additional_path.as_std_path());
+        }
+
+        builder.threads(
+            std::thread::available_parallelism()
+                .map_or(1, std::num::NonZeroUsize::get)
+                .min(12),
+        );
+
+        builder.build_parallel().run(|| {
+            let mut visitor = visitor_builder.build();
+
+            Box::new(move |entry| {
+                match entry {
+                    Ok(entry) => {
+                        // SAFETY: A system path can't be a stdin path.
```

Comment:
This safety comment isn't clear to me, I assume because I'm missing some important context. How do we know that `entry` is a system path? What is a "stdin path"?